### PR TITLE
Support LDoc

### PR DESCRIPTION
**Pull request by ligurio**
Online: https://tarantool.github.io/expirationd/

Ticket: #79

### CODE REVIEW

File: expirationd.lua [hunk 1]
```diff
@@ -310,48 +306,129 @@ end
...
+--     to process tuples from each batch in a single transaction.
+-- @boolean[opt] options.force
+--     expirationd doesn't work on replica by default. With option `force` it
+--     should be possible to force work on replica.
+-- @number[opt] options.full_scan_delay
+--     Sleep time between full scans (in seconds). It is allowed to pass an FFI
+--     number: `1LL`, `1UL` etc. Default value is 1 sec.
+-- @number[opt] options.full_scan_time
+--     Time required for a full index scan (in seconds). It is allowed to pass
+--     an FFI number: 1LL, 1UL etc. Default value is 3600.
```
Comment (ArtDu) [inline expirationd.lua:341]: The time required for a full scan of the index, but what happens if the real full scan time is longer than the specified time, or vice versa, faster? How does this field affect the logic of the entire work of the task? The logic for interacting with this field and others is partially described in this ticket #38, but not very obvious.

File: expirationd.lua [hunk 2]
```diff
@@ -192,12 +183,17 @@ end
 -- Task management
 -- ------------------------------------------------------------------------- --
 
+--[[
+task:start() - Force start `task` (old `guardian_fiber` will be garbage
+               collected, so do `task:stop()` before or `task:restart()` instead)
```
Comment (ArtDu) [inline expirationd.lua:188]: I know that the documentation for task: api is not ready yet. But I will comment on what you have already written.
> old `guardian_fiber` will be garbage collected, so do` task:stop()` before or` task:restart()` instead

I did not understand the point, it also calls `task:stop` inside and waits for the previous task to stop.

File: expirationd.lua [hunk 3]
```diff
@@ -310,48 +306,129 @@ end
...
+-- @string name
+--     Task name.
+-- @string space_id
+--     Space to look in for expired tuples.
+-- @func is_tuple_expired
+--     Function, must accept tuple and return true/false (is tuple expired or
+--     not), receives `args`, `tuple` as arguments.
+-- @table[opt] options
+--     Table with named options.
+-- @table[opt] options.args
```
Comment (ArtDu) [inline expirationd.lua:327]: Perhaps it is worth putting, at the beginning, the parameters that have a stronger effect on the customization of the task. For example, the index is often searched for and asked in the public Tarantool chat

Comment (ligurio) [reply expirationd.lua:327]: Options sorted alphabetically order by name. It's allow to find parameters that have strong effect on the customization personally for you ;)

File: expirationd.lua (hunk 2 above)
Comment (ligurio) [reply expirationd.lua]: > I know that the documentation for task: api is not ready yet. 

This PR is not about changing API, but about documenting it right and precisely.

> But I will comment on what you have already written.

It's text that was written before in comments, I just formatted it as LDoc comment. However I'll check it this still valid.

Comment (ligurio) [reply expirationd.lua]: Removed.

File: expirationd.lua [hunk 4]
```diff
@@ -327,48 +390,201 @@ end
...
+-- - Process min(`space_length`, `tuples_per_iteration`) tuples at once.
+--
+-- - Sleep `tuples_per_iteration` × `full_scan_time` / `space_length` (but not beyond 1 second).
+--
+-- - Repeat 1-2 until the whole space will be traversed.
+--
+-- - Sleep 1 second.
+--
+-- - Repeat 1-4.
+--
```
Comment (ArtDu) [inline expirationd.lua:414]: Maybe use numbers instead of dots. It's clearer with repeat 1-2/1-4?
Now:
image

Numbers:
image

File: doc/ldoc/assets/ldoc.ltp [hunk 5]
```diff
@@ -0,0 +1,420 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Comment (olegrok) [inline doc/ldoc/assets/ldoc.ltp:1]: Do we really need to push this generated code? I guess CI should push it github.io or at least we should have protection when we change something and don't update a doc.

Comment (ligurio) [reply doc/ldoc/assets/ldoc.ltp:1]: it's a template used by ldoc:

```
sergeyb@pony:~/sources/MRG/expirationd$ make apidoc 
.rocks/bin/ldoc -c /home/sergeyb/sources/MRG/expirationd/doc/ldoc/config.ld \
             -d /home/sergeyb/sources/MRG/expirationd/doc/apidoc/ expirationd.lua
reading configuration from /home/sergeyb/sources/MRG/expirationd/doc/ldoc/config.ld
template not found at 'doc/ldoc/assets' Use -l to specify directory containing ldoc.ltp
make: *** [Makefile:19: apidoc] Error 255
```

File: expirationd.lua (hunk 4 above)
Comment (ligurio) [reply expirationd.lua]: Sure, fixed.

```lua
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -402,15 +402,16 @@ end
 --
 -- How expirationd works in general:
 --
--- - Process min(`space_length`, `tuples_per_iteration`) tuples at once.
+-- 1. Process min(`space_length`, `tuples_per_iteration`) tuples at once.
 --
--- - Sleep `tuples_per_iteration` × `full_scan_time` / `space_length` (but not beyond 1 second).
+-- 2. Sleep `tuples_per_iteration` × `full_scan_time` / `space_length` (but not
+--    beyond 1 second).
 --
--- - Repeat 1-2 until the whole space will be traversed.
+-- 3. Repeat 1-2 until the whole space will be traversed.
 --
--- - Sleep 1 second.
+-- 4. Sleep 1 second.
 --
--- - Repeat 1-4.
+-- 5. Repeat 1-4.
 --
 --
 -- @string name
```

File: expirationd.lua [hunk 6]
```diff
@@ -327,48 +390,202 @@ end
...
+-- ```
+-- local function is_tuple_expired(args, tuple)
+--     local tuple_expire_time = get_field(tuple, args.field_no)
+--     if type(tuple_expire_time) ~= 'number' then
+--         return true
+--     end
+--     local current_time = fiber.time()
+--
+--     return current_time >= tuple_expire_time
+-- end
```
Comment (olegrok) [inline expirationd.lua:436]: image

looks broken

Comment (olegrok) [reply expirationd.lua:436]: image

functions below looks fine.

File: Makefile [hunk 7]
```diff
@@ -1,10 +1,24 @@
+# This way everything works as expected ever for
+# `make -C /path/to/project` or
+# `make -f /path/to/project/Makefile`.
+MAKEFILE_PATH := $(abspath $(lastword $(MAKEFILE_LIST)))
+PROJECT_DIR := $(patsubst %/,%,$(dir $(MAKEFILE_PATH)))
+LUACOV_REPORT := $(PROJECT_DIR)/luacov.report.out
```
Comment (Totktonada) [inline Makefile:6]: Luacov seems irrelevant here.

File: expirationd.lua [hunk 8]
```diff
@@ -327,48 +390,202 @@ end
 -- ========================================================================= --
 -- Expiration daemon management functions
 -- ========================================================================= --
+--
+-- {{{ Module methods
+--
+
+--- Module methods
+--
+-- @section Functions
```
Comment (Totktonada) [inline expirationd.lua:399]: In my understanding a method accepts an argument like `self` or `this`. So I think that 'functions' would be more appropriate term for those pieces of code.

File: expirationd.lua [hunk 9]
```diff
@@ -327,48 +390,202 @@ end
...
+-- 1. Process min(`space_length`, `tuples_per_iteration`) tuples at once.
+--
+-- 2. Sleep `tuples_per_iteration` × `full_scan_time` / `space_length` (but not
+--    beyond 1 second).
+--
+-- 3. Repeat 1-2 until the whole space will be traversed.
+--
+-- 4. Sleep 1 second.
+--
+-- 5. Repeat 1-4.
```
Comment (Totktonada) [inline expirationd.lua:414]: 1. Looks relevant to #38.
2. It is rendered as unnumbered list on https://tarantool.github.io/expirationd/, but when I generated it locally, the list is numbered. I guess the GitHub Pages webpage contains not a last state.

File: expirationd.lua [hunk 10]
```diff
@@ -327,48 +390,202 @@ end
...
+--
+--
+-- @string name
+--     Task name.
+-- @string space_id
+--     Space to look in for expired tuples.
```
Comment (Totktonada) [inline expirationd.lua:420]: A string or numerical ID, correct?

File: expirationd.lua [hunk 11]
```diff
@@ -327,48 +390,202 @@ end
...
+--     expired or not), receives `args` and `tuple` as arguments.
+--
+-- Example of function:
+--
+-- ```
+-- local function is_tuple_expired(args, tuple)
+--     local tuple_expire_time = get_field(tuple, args.field_no)
+--     if type(tuple_expire_time) ~= 'number' then
+--         return true
+--     end
```
Comment (Totktonada) [inline expirationd.lua:432]: I don't spot the idea of this check. If there is no expire information within a tuple, expire it anyway?

File: expirationd.lua [hunk 12]
```diff
@@ -327,48 +390,202 @@ end
...
+--     Table with named options.
+-- @table[opt] options.args
+--     Passed to `is_tuple_expired()` and `process_expired_tuple()` as
+--     an additional context.
+-- @boolean[opt] options.atomic_iteration
+--     False (default) to process each tuple as a single transaction and true
+--     to process tuples from each batch in a single transaction.
+-- @boolean[opt] options.force
+--     expirationd doesn't work on replica by default. With option `force` it
+--     should be possible to force work on replica.
```
Comment (Totktonada) [inline expirationd.lua:449]: It deserves a top level note (after this we can say that #14 is resolved). I heard 'why expirationd does not work' so many times...

I would also clarify what exactly 'replica' means in this context.

File: expirationd.lua [hunk 13]
```diff
@@ -327,48 +390,202 @@ end
...
+-- @number[opt] options.full_scan_delay
+--     Sleep time between full scans (in seconds). It is allowed to pass an FFI
+--     number: `1LL`, `1UL` etc. Default value is 1 sec.
```
Comment (Totktonada) [inline expirationd.lua:452]: Typo: 1UL -> 1ULL.

File: expirationd.lua [hunk 14]
```diff
@@ -327,48 +390,202 @@ end
...
+-- @number[opt] options.full_scan_time
+--     Time required for a full index scan (in seconds). It is allowed to pass
+--     an FFI number: `1LL`, `1UL` etc. `full_scan_time` used for calculation
```
Comment (Totktonada) [inline expirationd.lua:455]: Typo: 1UL -> 1ULL.

File: expirationd.lua [hunk 15]
```diff
@@ -327,48 +390,202 @@ end
...
+-- Example of function:
+--
+-- ```
+-- local function iterate_with()
+--     return true
+-- end
+-- ```
+-- @number[opt] options.iteration_delay
+--     Max sleep time between iterations (in seconds). It is allowed to pass
+--     an FFI number: `1LL`, `1UL` etc. Default value is 1 sec.
```
Comment (Totktonada) [inline expirationd.lua:477]: Typo: 1UL -> 1ULL.

File: expirationd.lua [hunk 16]
```diff
@@ -327,48 +390,202 @@ end
...
+--     Function which returns an iterator object which provides tuples to
+--     check, considering the `start_key`, `process_while` and other options.
+--     There's a default function which can be overriden with this parameter.
+--
+-- Example of function:
+--
+-- ```
+-- local function iterate_with()
+--     return true
+-- end
```
Comment (Totktonada) [inline expirationd.lua:473]: Just above:

> Function, which returns an iterator object \<...\>

But this function does not return an iterator object (`gen`, `param` `state` triplet, where `gen` is callable).

Comment (Totktonada) [reply expirationd.lua:473]: And I don't see any information about arguments, which is passed to the function.

File: expirationd.lua [hunk 17]
```diff
@@ -327,48 +390,202 @@ end
...
+--     Applied to expired tuples, receives `space_id`, `args`, `tuple` as
+--     arguments. When `process_expired_tuple` is not passed (or `nil` passed),
+--     tuples are removed.
+--
+-- Example of function:
+--
+-- ```
+-- local function put_tuple_to_archive(space_id, args, tuple)
+--     box.space[space_id]:delete{tuple[1]}
+--     local email = get_field(tuple, 2)
```
Comment (Totktonada) [inline expirationd.lua:516]: `tuple[1]` is used just above, but then `get_field(tuple, 2)`. Why? What is the difference?

File: expirationd.lua [hunk 18]
```diff
@@ -327,48 +390,202 @@ end
...
+--
+-- Example of function:
+--
+-- ```
+-- local function process_while()
+--     return false
+-- end
+-- ```
+--
+-- @string[opt] options.start_key
```
Comment (Totktonada) [inline expirationd.lua:536]: A key may be of any time. I would also say that even if expirationd allows to pass a constant key there, there is no much sense to do so (just bad api design). Any real usage will pass a function there.

File: expirationd.lua [hunk 19]
```diff
@@ -327,48 +390,202 @@ end
...
+--     Start iterating from the tuple with this index value. Or when iterator
+--     is 'EQ', iterate over tuples with this index value. Must be a value of
+--     the same data type as the index field or fields, or a function which
+--     returns such value. If omitted or nil, all tuples will be checked.
+-- @number[opt] options.tuples_per_iteration
+--     Number of tuples to check in one batch (iteration). It is allowed to
+--     pass an FFI number: `1LL`, `1UL` etc. Default value is 1024.
```
Comment (Totktonada) [inline expirationd.lua:543]: Typo: 1UL -> 1ULL.

File: expirationd.lua [hunk 20]
```diff
@@ -327,48 +390,202 @@ end
...
+-- @number[opt] options.vinyl_assumed_space_len_factor
+--     Factor for recalculation of vinyl space size. Vinyl space size can't be
+--     counted (since many operations, `upsert` for example, are applied when
+--     you address some data), so you should count (approximate space size)
+--     tuples with the first start. `vinyl_assumed_space_len` is approximate
+--     count for first run and `vinyl_assumed_space_len_factor` for next
+--     milestone (after we've reached next milestone is `*` and so on). It is
+--     allowed to pass an FFI number: `1LL`, `1UL` etc. Default value is 2.
```
Comment (Totktonada) [inline expirationd.lua:551]: Same, 1UL -> 1ULL.

File: expirationd.lua [hunk 21]
```diff
@@ -327,48 +390,202 @@ end
...
+-- @number[opt] options.vinyl_assumed_space_len
+--     Assumed size of vinyl space (in the first iteration).
+--     Vinyl space size can't be counted (since many operations, `upsert` for
+--     example, are applied when you address some data), so you should count
+--     (approximate space size) tuples with the first start.
+--     `vinyl_assumed_space_len` is approximate count for first run and
+--     `vinyl_assumed_space_len_factor` for next milestone (after we've reached
+--     next milestone is `*` and so on). It is allowed to pass an FFI number:
+--     `1LL`, `1UL` etc. Default value is 10^7.
```
Comment (Totktonada) [inline expirationd.lua:560]: Same, 1UL -> 1ULL.

File: expirationd.lua [hunk 22]
```diff
@@ -327,48 +390,202 @@ end
...
+--
+-- @treturn task_obj
```
Comment (Totktonada) [inline expirationd.lua:562]: The link goes to the task function. Is it correct? I would expect that it'll go to the list of task object methods.

File: Makefile (hunk 7 above)
Comment (ligurio) [reply Makefile:6]: Right.

```
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,6 @@
 # `make -f /path/to/project/Makefile`.
 MAKEFILE_PATH := $(abspath $(lastword $(MAKEFILE_LIST)))
 PROJECT_DIR := $(patsubst %/,%,$(dir $(MAKEFILE_PATH)))
-LUACOV_REPORT := $(PROJECT_DIR)/luacov.report.out
 
 CLEANUP_FILES  = tarantool.log
 CLEANUP_FILES += *.xlog*
```

File: expirationd.lua (hunk 8 above)
Comment (ligurio) [reply expirationd.lua:399]: Fixed.

```lua
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -391,10 +391,10 @@ end
 -- Expiration daemon management functions
 -- ========================================================================= --
 --
--- {{{ Module methods
+-- {{{ Functions
 --
 
---- Module methods
+--- Functions
 --
 -- @section Functions
 
@@ -924,4 +924,4 @@ return {
     show_task_list = show_task_list_obsolete,
 }
 
--- }}} Module methods
+-- }}} Functions
```

File: expirationd.lua (hunk 10 above)
Comment (ligurio) [reply expirationd.lua:420]: Right, fixed it.

File: expirationd.lua (hunk 11 above)
Comment (ligurio) [reply expirationd.lua:432]: Fixed it. (copy-pasted from test)

```lua
 -- local function is_tuple_expired(args, tuple)
 --     local tuple_expire_time = get_field(tuple, args.field_no)
---     if type(tuple_expire_time) ~= 'number' then
---         return true
---     end
 --     local current_time = fiber.time()
---
 --     return current_time >= tuple_expire_time
 -- end
```

File: expirationd.lua (hunk 13 above)
Comment (ligurio) [reply expirationd.lua:452]: Fixed.

File: expirationd.lua (hunk 15 above)
Comment (ligurio) [reply expirationd.lua:477]: Fixed.

File: expirationd.lua (hunk 14 above)
Comment (ligurio) [reply expirationd.lua:455]: Fixed.

File: expirationd.lua (hunk 18 above)
Comment (ligurio) [reply expirationd.lua:536]: > A key may be of any time.

Type?


True. Below is a piece of code to process `start_key` argument:
```lua
    -- check start_key                                                           
    if options.start_key ~= nil or options.start_key == box.NULL then            
        if type(options.start_key) == "function" then                            
            task.start_key = function() return options.start_key() end           
        else                                                                     
            task.start_key = function() return options.start_key end             
        end                                                                      
    end      
```

Fixed a type of `start_key` in LDoc comment.

P.S. type of `start_key` in checks() is correct:
```lua
local function expirationd_run_task(name, space_id, is_tuple_expired, options)   
    checks('string', 'number|string', 'function', {                              
        args = '?',                                                              
        ...                                         
        start_key = '?',                                                         
        tuples_per_iteration = '?number|cdata',                                  
        vinyl_assumed_space_len_factor = '?number|cdata',                        
        vinyl_assumed_space_len = '?number|cdata',                               
    })                 
```

File: expirationd.lua (hunk 20 above)
Comment (ligurio) [reply expirationd.lua:551]: Fixed.

File: expirationd.lua (hunk 21 above)
Comment (ligurio) [reply expirationd.lua:560]: Fixed.

File: expirationd.lua (hunk 19 above)
Comment (ligurio) [reply expirationd.lua:543]: Fixed,

File: expirationd.lua (hunk 9 above)
Comment (ligurio) [reply expirationd.lua:414]: 2. Oleg already spotted it and I haven't update HTML version after fix.
Pushed latest version to https://tarantool.github.io/expirationd/.

File: expirationd.lua (hunk 16 above)
Comment (ligurio) [reply expirationd.lua:473]: Removed example and instead described default function:
```lua
@@ -463,18 +460,25 @@ end
 -- @func[opt] options.iterate_with
 --     Function which returns an iterator object which provides tuples to
 --     check, considering the `start_key`, `process_while` and other options.
---     There's a default function which can be overriden with this parameter.
---
--- Example of function:
+--     When option is nil default function is used. Default function returns
+--     iterator returned by [index_object:pairs()][1], where `index` is a
+--     primary index or index that specified with argument `options.index`:
 --
 -- ```
--- local function iterate_with()
---     return true
--- end
+--  index:pairs(option.start_key(), {
+--     iterator = task.iterator_type
+--  }):take_while(
+--         function()
+--             return option.process_while()
+--         end
+--    )
 -- ```
+--
+--     [1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/pairs/.
+--
```

> And I don't see any information about arguments, which is passed to the function.

it is a question how to describe it, because `task` is an internal object. Perhaps we need to change arguments passed to `iterate_with` to make possible to describe it in documentation. What do you think?

File: expirationd.lua (hunk 17 above)
Comment (ligurio) [reply expirationd.lua:516]: Fixed.
```lua
@@ -513,7 +517,7 @@ end
 -- ```
 -- local function put_tuple_to_archive(space_id, args, tuple)
 --     box.space[space_id]:delete{tuple[1]}
---     local email = get_field(tuple, 2)
+--     local email = tuple[2]
 --     if args.archive_space_id ~= nil and email ~= nil then
 --         box.space[args.archive_space_id]:replace{email, fiber.time()}
 --     end
```


It was copy-pasted from TAP tests and `get_field` handle cases when tuple is `nil` field_no is bigger than `#tuple`:
```lua
-- get field                                                                     
local function get_field(tuple, field_no)                                        
    if tuple == nil then                                                         
        return nil                                                               
    end                                                                          
                                                                                 
    if #tuple < field_no then                                                    
        return nil                                                               
    end                                                                          
                                                                                 
    return tuple[field_no]                                                       
end                 
```

We don't need get_field< so I removed it.

File: expirationd.lua (hunk 22 above)
Comment (ligurio) [reply expirationd.lua:562]: I keep it as is because I don't get how to fix it in LDoc.

File: expirationd.lua (hunk 6 above)
Comment (ligurio) [reply expirationd.lua]: Fixed

Screenshot from 2021-09-27 17-11-07

File: expirationd.lua (hunk 12 above)
Comment (Totktonada) [reply expirationd.lua]: It's good, but I meant a note somewhere at the module level. Maybe as part of usage example. If one don't know that there is a trick around `force` option to make expirationd work on a master with read only upstreams, (s)he will not know where to find the trick. It is the key idea of #14 (the issue appears despite the fact that the force option was mentioned in the readme).

File: expirationd.lua (hunk 16 above)
Comment (Totktonada) [reply expirationd.lua]: Okay.

File: expirationd.lua (hunk 12 above)
Comment (Totktonada) [reply expirationd.lua]: Okay, thanks! (I would place it somewhere in the documentation, but README is okay for me too.)
